app/plugins/middleware.py

```diff
@@ -0,0 +1,191 @@
+    def __init__(self, app: Any, exclude_paths: Optional[list[str]] = None) -> None:
+        super().__init__(app)
+        self.exclude_paths = exclude_paths or [
+            "/docs",
+            "/redoc",
+            "/openapi.json",
+            "/",
+            "/api/v1/auth",
+        ]
+
+    async def dispatch(self, request: Request, call_next: Callable[[Request], Awaitable[Response]]) -> Response:
+        if self._is_excluded_path(request.url.path):
+            response: Response = await call_next(request)
+            return response
+
+        plugin_name = self._get_route_plugin_name(request)
+        if not plugin_name:
+            response = await call_next(request)
+            return response
+
+        user = getattr(request.state, "user", None)
+        if not user:
+            response = await call_next(request)
+            return response
+
+        has_access = await self._check_plugin_access(user.id, plugin_name)
+        if not has_access:
+            logger.warning(
+                f"User {user.id} attempted to access disabled plugin '{plugin_name}' at path {request.url.path}"
+            )
+            return JSONResponse(
+                status_code=status.HTTP_403_FORBIDDEN,
+                content={
+                    "detail": f"Access to plugin '{plugin_name}' is disabled for your account. "
+                    f"Please enable the plugin in your settings."
+                },
+            )
+
+        response = await call_next(request)
+        return response
+
+    def _is_excluded_path(self, path: str) -> bool:
+        for excluded in self.exclude_paths:
+            if path.startswith(excluded):
+                return True
+        return False
+
+    def _get_route_plugin_name(self, request: Request) -> Optional[str]:
+        for route in request.app.routes:
+            match, _ = route.matches(request.scope)
+            if match == Match.FULL:
+                if hasattr(route, "endpoint"):
+                    endpoint = route.endpoint
+                    if hasattr(endpoint, "__plugin_name__"):
+                        plugin_name = getattr(endpoint, "__plugin_name__")
+                        if isinstance(plugin_name, str):
+                            return plugin_name
+                    if hasattr(route, "tags") and route.tags:
+                        for tag in route.tags:
+                            if isinstance(tag, str) and tag.startswith("plugin:"):
+                                return tag.replace("plugin:", "")
+                break
+        return None
+
+    async def _check_plugin_access(self, user_id: int, plugin_name: str) -> bool:
+        try:
+            async for session in get_async_session():
+                return await _check_plugin_access_impl(user_id, plugin_name, session, check_system_enabled=True)
+            # Fallback if session generator doesn't yield
+            return False
+        except SQLAlchemyError as e:
```

Comment:
`SQLAlchemyError` is too broad

Reply:
Created new issue for it:

https://github.com/edly-io/sparkth/issues/123